src/cryptography.rs

```diff
@@ -73,12 +77,35 @@ fn check_account_auth(
     }
 }
 
-pub fn check_auth(e: &Env, auth: KeyedAuthorization, domain: Domain, parameters: EnvVal) {
+pub fn check_auth<T>(
```

Comment:
This approach makes me extremely nervous. Specifically, I don't like the fact that the `nonce_key` isn't derived from the `auth` at this point. It's possible to call with the wrong `nonce_key`. This applies for all of these functions.

Is there anything we could do to avoid this design? For example, one (not necessarily good) idea would be to do a newtype around `KeyedAuthorization` and have that newtype implement a specific trait that converts it to a nonce-key... or even just gets the nonce directly.

Reply:
Something like that would work. I went this route initially because I wanted to handle the `Authorization` use case as well and this was easier, but it's too general. I think I'll hold off on this PR and revisit once we refactor authorization.